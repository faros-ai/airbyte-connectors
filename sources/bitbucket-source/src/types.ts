import {AirbyteConfig} from 'faros-airbyte-cdk';

export interface BitbucketConfig extends AirbyteConfig {
  readonly api_url?: string;
  readonly username?: string;
  readonly password?: string;
  readonly token?: string;
  readonly workspaces?: ReadonlyArray<string>;
  readonly excluded_workspaces?: ReadonlyArray<string>;
  readonly repositories?: ReadonlyArray<string>;
  readonly excluded_repositories?: ReadonlyArray<string>;
  readonly page_size?: number;
  readonly bucket_id?: number;
  readonly bucket_total?: number;
  readonly cutoff_days?: number;
  readonly start_date?: string;
  readonly end_date?: string;
  // startDate and endDate are calculated from start_date, end_date, and cutoff_days
  startDate?: Date;
  endDate?: Date;
}
