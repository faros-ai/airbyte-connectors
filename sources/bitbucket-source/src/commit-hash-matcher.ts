export class CommitHashMatcher {
  private sortedHashes: string[];

  constructor(hashes: Set<string>) {
    this.sortedHashes = Array.from(hashes).sort();
  }

  match(shortHash: string): string | null {
    let left = 0;
    let right = this.sortedHashes.length - 1;

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      const midHash = this.sortedHashes[mid];

      if (midHash.startsWith(shortHash)) {
        return midHash;
      } else if (midHash < shortHash) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return null;
  }
}
