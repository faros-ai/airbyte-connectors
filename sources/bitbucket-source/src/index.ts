import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Bitbucket} from './bitbucket/bitbucket';
import {BitbucketConfig} from './bitbucket/types';
import {
  Branches,
  Commits,
  Deployments,
  Issues,
  Pipelines,
  PipelineSteps,
  PullRequestActivities,
  PullRequests,
  Repositories,
  Workspaces,
  WorkspaceUsers,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new BitbucketSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class BitbucketSource extends AirbyteSourceBase<BitbucketConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    /* eslint-disable-next-line @typescript-eslint/no-var-requires */
    return new AirbyteSpec(require('../resources/spec.json'));
  }

  async checkConnection(config: BitbucketConfig): Promise<[boolean, VError]> {
    try {
      const bitbucket = Bitbucket.instance(
        config as BitbucketConfig,
        this.logger
      );
      await bitbucket.checkConnection();
    } catch (error: any) {
      return [false, error];
    }
    return [true, undefined];
  }

  streams(config: BitbucketConfig): AirbyteStreamBase[] {
    const pipelines = new Pipelines(config, this.logger);
    const pullRequests = new PullRequests(config, this.logger);
    return [
      new Branches(config, this.logger),
      new Commits(config, this.logger),
      new Deployments(config, this.logger),
      new Issues(config, this.logger),
      pipelines,
      new PipelineSteps(config, pipelines, this.logger),
      pullRequests,
      new PullRequestActivities(config, pullRequests, this.logger),
      new Repositories(config, this.logger),
      new WorkspaceUsers(config, this.logger),
      new Workspaces(config, this.logger),
    ];
  }
}
