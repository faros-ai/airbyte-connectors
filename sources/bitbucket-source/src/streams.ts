import {
  AirbyteLogger,
  AirbyteStreamBase,
  StreamKey,
  SyncMode,
} from 'faros-airbyte-cdk';
import {Dictionary} from 'ts-essentials';

import {createClient} from './bitbucket';
import {BitbucketConfig, Branch, Workspace} from './types';

export class BitbucketBranches extends AirbyteStreamBase {
  constructor(
    readonly config: BitbucketConfig,
    readonly repositories: string[],
    readonly logger: AirbyteLogger
  ) {
    super(logger);
  }

  getJsonSchema(): Dictionary<any, string> {
    return require('../resources/schemas/branches.json');
  }

  get primaryKey(): StreamKey {
    return [];
  }

  async *streamSlices(
    syncMode: SyncMode,
    cursorField?: string[],
    streamState?: Dictionary<any>
  ): AsyncGenerator<Dictionary<any> | undefined> {
    for (const repository of this.repositories) {
      yield {repository};
    }
  }

  async *readRecords(
    syncMode: SyncMode,
    cursorField?: string[],
    streamSlice?: Dictionary<any, string>,
    streamState?: Dictionary<any, string>
  ): AsyncGenerator<Branch, any, unknown> {
    const [client, errorMessage] = await createClient(this.config);
    if (!client) {
      this.logger.error(errorMessage);
      return undefined;
    }

    const repo_slug = streamSlice.repository;
    yield* client.getBranches(this.config.workspace, repo_slug);
  }
}

export class BitbucketRepositories extends AirbyteStreamBase {
  constructor(
    readonly config: BitbucketConfig,
    readonly repositories: string[],
    readonly logger: AirbyteLogger
  ) {
    super(logger);
  }

  getJsonSchema(): Dictionary<any, string> {
    return require('../resources/schemas/repositories.json');
  }

  get primaryKey(): StreamKey {
    return ['uuid'];
  }

  async *streamSlices(
    syncMode: SyncMode,
    cursorField?: string[],
    streamState?: Dictionary<any>
  ): AsyncGenerator<Dictionary<any> | undefined> {
    for (const repository of this.repositories) {
      yield {repository};
    }
  }

  async *readRecords(
    syncMode: SyncMode,
    cursorField?: string[],
    streamSlice?: Dictionary<any, string>,
    streamState?: Dictionary<any, string>
  ): AsyncGenerator<Dictionary<any, string>, any, unknown> {
    const [client, errorMessage] = await createClient(this.config);
    if (!client) {
      this.logger.error(errorMessage);
      return undefined;
    }

    const repo_slug = streamSlice.repository;
    yield await client.getRepository(this.config.workspace, repo_slug);
  }
}

export class BitbucketWorkspaces extends AirbyteStreamBase {
  constructor(readonly config: BitbucketConfig, logger: AirbyteLogger) {
    super(logger);
  }

  getJsonSchema(): Dictionary<any, string> {
    return require('../resources/schemas/workspaces.json');
  }

  get primaryKey(): StreamKey {
    return ['uuid'];
  }

  async *readRecords(
    syncMode: SyncMode,
    cursorField?: string[],
    streamSlice?: Workspace,
    streamState?: Dictionary<any, string>
  ): AsyncGenerator<Workspace, any, unknown> {
    const [client, errorMessage] = await createClient(this.config);
    if (!client) {
      this.logger.error(errorMessage);
      return undefined;
    }

    yield await client.getWorkspace(this.config.workspace);
  }
}
