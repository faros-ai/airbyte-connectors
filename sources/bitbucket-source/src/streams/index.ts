import {Branches} from './branches';
import {Commits} from './commits';
import {Deployments} from './deployments';
import {Issues} from './issues';
import {PipelineSteps} from './pipeline_steps';
import {Pipelines} from './pipelines';
import {PullRequestActivities} from './pull_request_activities';
import {PullRequests} from './pull_requests';
import {Repositories} from './repositories';
import {WorkspaceUsers} from './workspace_users';
import {Workspaces} from './workspaces';

export {
  Branches,
  Commits,
  Deployments,
  Issues,
  PipelineSteps,
  Pipelines,
  PullRequestActivities,
  PullRequests,
  Repositories,
  WorkspaceUsers,
  Workspaces,
};
