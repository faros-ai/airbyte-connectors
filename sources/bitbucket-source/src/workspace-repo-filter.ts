import {AirbyteLogger} from 'faros-airbyte-cdk';
import {Repository} from 'faros-airbyte-common/bitbucket';
import {collectReposByOrg} from 'faros-airbyte-common/common';
import {Memoize} from 'typescript-memoize';
import VError from 'verror';

import {Bitbucket} from './bitbucket';
import {BitbucketConfig} from './types';

type FilterConfig = {
  workspaces?: Set<string>;
  excludedWorkspaces?: Set<string>;
  reposByWorkspace: Map<string, Set<string>>;
  excludedReposByWorkspace: Map<string, Set<string>>;
};

export class WorkspaceRepoFilter {
  private readonly filterConfig: FilterConfig;
  private workspaces?: Set<string>;
  private reposByWorkspace: Map<string, Map<string, Repository>> = new Map();

  constructor(
    private readonly config: BitbucketConfig,
    private readonly logger: AirbyteLogger
  ) {
    const {workspaces, repositories, excluded_repositories} = this.config;
    let {excluded_workspaces} = this.config;

    if (workspaces?.length && excluded_workspaces?.length) {
      this.logger.warn(
        'Both workspaces and excluded_workspaces are specified, excluded_workspaces will be ignored.'
      );
      excluded_workspaces = undefined;
    }

    const reposByWorkspace = new Map<string, Set<string>>();
    if (repositories?.length) {
      collectReposByOrg(reposByWorkspace, repositories);
    }
    const excludedReposByWorkspace = new Map<string, Set<string>>();
    if (excluded_repositories?.length) {
      collectReposByOrg(excludedReposByWorkspace, excluded_repositories);
    }

    for (const workspace of reposByWorkspace.keys()) {
      if (excludedReposByWorkspace.has(workspace)) {
        this.logger.warn(
          `Both repositories and excluded_repositories are specified for workspace ${workspace}, excluded_repositories for workspace ${workspace} will be ignored.`
        );
        excludedReposByWorkspace.delete(workspace);
      }
    }

    this.filterConfig = {
      workspaces: workspaces?.length ? new Set(workspaces) : undefined,
      excludedWorkspaces: excluded_workspaces?.length
        ? new Set(excluded_workspaces)
        : undefined,
      reposByWorkspace,
      excludedReposByWorkspace,
    };
  }

  @Memoize()
  async getWorkspaces(): Promise<ReadonlyArray<string>> {
    if (!this.workspaces) {
      const workspaces = new Set<string>();
      const bitbucket = Bitbucket.instance(this.config, this.logger);
      const visibleWorkspaces = await bitbucket.getWorkspaceIds();
      if (!visibleWorkspaces.length) {
        this.logger.warn('No visible workspaces found');
      }
      if (!this.filterConfig.workspaces) {
        visibleWorkspaces.forEach((org) => {
          if (!this.filterConfig.excludedWorkspaces?.has(org)) {
            workspaces.add(org);
          }
        });
      } else {
        this.filterConfig.workspaces.forEach((workspace) => {
          if (
            visibleWorkspaces.length &&
            visibleWorkspaces.includes(workspace)
          ) {
            workspaces.add(workspace);
          }
        });
      }
      this.workspaces = workspaces;
    }
    return Array.from(this.workspaces);
  }

  @Memoize()
  async getRepositories(workspace: string): Promise<ReadonlyArray<Repository>> {
    if (!this.reposByWorkspace.has(workspace)) {
      const repos = new Map<string, Repository>();
      const bitbucket = Bitbucket.instance(this.config, this.logger);
      const visibleRepos = await bitbucket.getRepositories(workspace);
      if (!visibleRepos.length) {
        this.logger.warn(
          `No visible repositories found for workspace ${workspace}`
        );
      }
      if (!this.filterConfig.reposByWorkspace.has(workspace)) {
        visibleRepos.forEach((repo) => {
          if (
            !this.filterConfig.excludedReposByWorkspace
              .get(workspace)
              ?.has(repo.slug)
          ) {
            repos.set(repo.slug, repo);
          }
        });
      } else {
        this.filterConfig.reposByWorkspace
          .get(workspace)
          .forEach((repoName) => {
            const repo = visibleRepos.find((r) => r.slug === repoName);
            if (!repo) {
              this.logger.warn(
                `Skipping not found repository ${workspace}/${repoName}`
              );
              return;
            }
            repos.set(repo.slug, repo);
          });
      }
      this.reposByWorkspace.set(workspace, repos);
    }
    return Array.from(this.reposByWorkspace.get(workspace).values());
  }

  getRepository(org: string, name: string): Repository {
    const repo = this.reposByWorkspace.get(org)?.get(name);
    if (!repo) {
      throw new VError('Repository not found: %s/%s', org, name);
    }
    return repo;
  }
}
