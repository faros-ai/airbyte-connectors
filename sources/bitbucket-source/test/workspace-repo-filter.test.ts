import {
  AirbyteLogger,
  AirbyteLogLevel,
  readTestResourceAsJSON,
} from 'faros-airbyte-cdk';

import {Bitbucket} from '../src/bitbucket';
import {BitbucketConfig} from '../src/types';
import {WorkspaceRepoFilter} from '../src/workspace-repo-filter';

const bitbucketInstance = jest.fn().mockImplementation(() => {
  return new Bitbucket(
    {
      repositories: {
        list: jest.fn().mockResolvedValue({
          data: {
            values: [{slug: 'repo-1'}, {slug: 'repo-2'}, {slug: 'repo-3'}],
          },
        }),
      },
      workspaces: {
        getWorkspaces: jest.fn().mockResolvedValue({
          data: {
            values: [
              {slug: 'workspace-1'},
              {slug: 'workspace-2'},
              {slug: 'workspace-3'},
            ],
          },
        }),
      },
      hasNextPage: jest.fn(),
    } as any,
    100,
    1,
    1,
    new AirbyteLogger()
  );
});

describe('workspaceRepoFilter', () => {
  const logger = new AirbyteLogger(AirbyteLogLevel.DEBUG);
  const config: BitbucketConfig = readTestResourceAsJSON('config.json');

  beforeEach(() => {
    Bitbucket.instance = bitbucketInstance;
  });

  test('getWorkspaces - all - no list', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(config, logger);
    const workspaces = await workspaceRepoFilter.getWorkspaces();
    expect(workspaces).toMatchSnapshot();
  });

  test('getWorkspaces - all - empty list', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        workspaces: [],
      },
      logger
    );
    const workspaces = await workspaceRepoFilter.getWorkspaces();
    expect(workspaces).toMatchSnapshot();
  });

  test('getWorkspaces - specific workspaces included', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        workspaces: ['workspace-1', 'workspace-2'],
      },
      logger
    );
    const workspaces = await workspaceRepoFilter.getWorkspaces();
    expect(workspaces).toMatchSnapshot();
  });

  test('getWorkspaces - specific workspaces excluded', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        excluded_workspaces: ['workspace-1', 'workspace-2'],
      },
      logger
    );
    const workspaces = await workspaceRepoFilter.getWorkspaces();
    expect(workspaces).toMatchSnapshot();
  });

  test('getRepositories - all - no list', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(config, logger);
    const repositories = await getAllRepositories(workspaceRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - all - empty list', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        repositories: [],
      },
      logger
    );
    const repositories = await getAllRepositories(workspaceRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - specific repositories included', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        repositories: ['workspace-1/repo-1', 'workspace-2/repo-2'],
      },
      logger
    );
    const repositories = await getAllRepositories(workspaceRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - specific repositories excluded', async () => {
    const workspaceRepoFilter = new WorkspaceRepoFilter(
      {
        ...config,
        excluded_repositories: ['workspace-1/repo-1', 'workspace-2/repo-2'],
      },
      logger
    );
    const repositories = await getAllRepositories(workspaceRepoFilter);
    expect(repositories).toMatchSnapshot();
  });
});

const getAllRepositories = async (workspaceRepoFilter: WorkspaceRepoFilter) => {
  const repos = [
    ...(await workspaceRepoFilter.getRepositories('workspace-1')),
    ...(await workspaceRepoFilter.getRepositories('workspace-2')),
    ...(await workspaceRepoFilter.getRepositories('workspace-3')),
  ];
  return repos.map((repo) => ({slug: repo.slug, workspace: repo.workspace}));
};
