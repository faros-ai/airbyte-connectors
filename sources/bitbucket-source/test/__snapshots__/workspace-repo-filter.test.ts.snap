// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspaceRepoFilter getRepositories - all - empty list 1`] = `
[
  {
    "slug": "repo-1",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-3",
  },
]
`;

exports[`workspaceRepoFilter getRepositories - all - no list 1`] = `
[
  {
    "slug": "repo-1",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-3",
  },
]
`;

exports[`workspaceRepoFilter getRepositories - specific repositories excluded 1`] = `
[
  {
    "slug": "repo-2",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-3",
  },
]
`;

exports[`workspaceRepoFilter getRepositories - specific repositories included 1`] = `
[
  {
    "slug": "repo-1",
    "workspace": "workspace-1",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-2",
  },
  {
    "slug": "repo-1",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-2",
    "workspace": "workspace-3",
  },
  {
    "slug": "repo-3",
    "workspace": "workspace-3",
  },
]
`;

exports[`workspaceRepoFilter getWorkspaces - all - empty list 1`] = `
[
  "workspace-1",
  "workspace-2",
  "workspace-3",
]
`;

exports[`workspaceRepoFilter getWorkspaces - all - no list 1`] = `
[
  "workspace-1",
  "workspace-2",
  "workspace-3",
]
`;

exports[`workspaceRepoFilter getWorkspaces - specific workspaces excluded 1`] = `
[
  "workspace-3",
]
`;

exports[`workspaceRepoFilter getWorkspaces - specific workspaces included 1`] = `
[
  "workspace-1",
  "workspace-2",
]
`;
